Add new test type, collected yield.

CollinValley opened the pull request:
The commit messages should speak for themselves. But collected yield is a test where we actually collect the output from the router so that we can compare it to the expected output and produce a pass/fail case that is not determined solely by whether the runtime returns or not.  I imagine that at some point almost all of our tests will follow the collected yield model, but this is a good first step.  To support this I created a new generator and collector. I think they are suitably generic for now, we may need some more features down the road, but I don't know what that might be at this juncture. 

I also include a commit that reduces the verbosity of our test pieces, I figure we should really only be printing output on a test failure, so that the test output is not impossible to read. This might be contentious though so I welcome feedback on this move.

CollinValley committed:
packet_collectors.rs:  Impl Exhaustive Colllector

Exhaustive Collector is a packet collector that will grab every packet
out of a stream, but it is different from Exhaustive Drain in that it
writes the collected packet into the provided channel handle. This is
useful because the test may then inspect the channel to determine
whether the test was a pass for fail.

CollinValley committed:
packet_generators.rs: Impl PacketIntervalGenerator

This commit adds the PacketIntervalGenerator, a generator that
takes an interval and an object that implements Iterator, and pulls
packets from that object until it is exhausted, provided the packets
out of a Stream. The packets are produced on a fixed interval so that
the Stream sometimes returns NotReady,  to better test the router.

CollinValley committed:
lib.rs: Add Collected Yield Tests

Up unti this point, our testing suite did not have a way to assert that
the output of the router matched some sort of specification.  The
tests pass or fail simply based on whether all of the futures return.

This commit puts the use the PacketIntervalGenerator and the
ExhaustiveCollector to create a suite of tests that the assert that
the series of packets that were fed into the router match the
output of the router.

Run the suite with the command `cargo test collected_yield`

CollinValley committed:
Test: Reduce verbosity of successful tests

It isn't considered good form for tests to print when they are not
throwing some kind of error.  With all the prints in our testbed it can
be hard to read what passed and what failed. Made this better by
commenting out all the debugging print statements. I figure that
when there is a failed test it will be convienient to have those there
to be uncommented.  After this commit tests are much cleaner and
running `cargo test` is functional.

CollinValley committed:
AsyncElement: Fix some typos, small errors

We had a few non-breaking typos in the AsyncElement implementation,
1)  We were doing a blocking send instead to a non-blocking
try_send() on a channel. It should never fail so it passed tests.

2) Got the await_consumer, and await_provider switched in their
declaration in the new function for AsyncElementLink. But we got
it backwards twice so it worked out, fixes that mistake.

ddowl commented on src/utils/test/packet_collectors.rs:
```diff
@@ -33,3 +34,39 @@ impl<T: Debug> Future for ExhaustiveDrain<T> {
+}
+
+impl<T: Debug> ExhaustiveCollector<T> {
+    pub fn new(id: usize, stream: ElementStream<T>, packet_dump: crossbeam_channel::Sender<T>) -> Self {
```
In other places we've been typing by just `Sender<T>`. Could we import the whole `crossbeam::crossbeam_channel::Sender` please?

CollinValley replied:
Will clean this up.  I'm going to go with calls to `bounded` will be `crossbeam_channel::bounded`,  calls to Sender and Receiver will be `Sender` and `Receiver`.   Similar pattern for unbounded.  I think that balances readability with succinctness.

CollinValley replied:
Cleaned up as requested. Future PRs will also follow this convention.

ddowl commented on src/utils/test/packet_generators.rs:
```diff
@@ -52,3 +52,47 @@ impl Stream for LinearIntervalGenerator {
     }
 }
+
+/// Packet Interval Generator procduces a Stream of packets on a defined interval.AsMut
```
The "LineraInterval" of `LinearIntervalGenerator` was meant to indicate that some packets are generated at a _fixed_ interval, e.g. every 100ms. The name `PacketIntervalGenerator` doesn't really make sense in this way. Maybe we should rename it to `FixedIntervalGenerator`, since you could also interpret the "Interval" as the ranges of digits, but that's not the intention. Let's deprecate `LinearIntervalGenerator`, since all ranges can be expressed by passing `range.into_iter()` to a `FixedIntervalGenerator`.

ddowl replied:
We should also try to deprecate `LinearIntervalGenerator` since there's a lot of duplicated logic between the two.

CollinValley replied:
Leave deprecation for testing fixup issue.  I get the names are bad, but we shouldn't change too much in one PR

ddowl commented on src/utils/test/packet_collectors.rs:
```diff
@@ -33,3 +34,39 @@ impl<T: Debug> Future for ExhaustiveDrain<T> {
+    }
+}
+
+impl<T: Debug> Future for ExhaustiveCollector<T> {
```
I wish there was a way to reduce redundancy, since `ExhaustiveDrain` and `ExhaustiveCollector` are so similar.

Could `ExhaustiveDrain` just contain an `ExhaustiveCollector` that does nothing with its channel?

CollinValley replied:
Let's leave this for a separate PR.  We are going to drop many tests all together.

ddowl commented on src/lib.rs:
```diff
@@ -52,6 +52,26 @@ mod tests {
     }
 
+    #[test]
+    fn one_sync_element_collected_yield() {
```
I'm unsure why we're only adding more tests here instead of replacing all of the existing tests with the packet collection assertion. Why?

Also, when all of our tests pass/fail on assertions, the `collected_yield` suffix will be redundant.

CollinValley replied:
Yes, I think this is true. Can we clean this up later during the testing fixup. I was did not want to depreciate existing work until this PR is in.

ddowl commented on src/lib.rs:
```diff
@@ -25,7 +26,7 @@ mod tests {
 
         fn process(&mut self, packet: Self::Input) -> Self::Output {
-            println!("Got packet {} in element {}", packet, self.id);
+            //println!("Got packet {} in element {}", packet, self.id);
```
In any case, we shouldn't be committing commented code, but I don't think removing logging altogether is a good idea. It would be better if we could enable logging with via an environment variable, at least until we can enable `tokio_trace`.

I think something like [env_logger](https://docs.rs/env_logger/0.6.2/env_logger/) does the job quite well. That can be introduced in a separate commit. I'd be happy to look into that if you want.

CollinValley replied:
Yes that would be a good idea, leave in for now,  resolve in new commit.

ddowl commented on src/api/mod.rs:
```diff
@@ -77,14 +77,15 @@ pub struct AsyncElementLink< E: AsyncElement> {
 
 impl<E: AsyncElement> AsyncElementLink<E> {
     pub fn new(input_stream: ElementStream<E::Input>, element: E, queue_capacity: usize) -> Self {
+        assert!( queue_capacity <= 1000, format!("Async Element queue_capacity: {} > 1000", queue_capacity));
```
I think this commit is useful, but should it be in this PR?

CollinValley replied:
Hmm, that was probably added slightly by mistake.  Will fork that out somewhere else if at all possible.

CollinValley replied:
Removed this to be added back in on a later PR

ddowl reviewed (changes requested):
I like the main ideas of the generic generator and packet collector here. There are a few comments that we should address before merging.

CollinValley committed:
Collected Yield: Cleanup for PR

This commit resolves some of the issues brought up by the PR for
this commit stack.

1) It makes all calls to Sender and Reciever direct rather than
   through calling crossbeam_channel::... This is easier to read.

2) When calling for a bounded or unbounded channel, we use the call
   crossbeam_channel. Bounded and Unbounded without context in the
   codebase cause a lot of confusion, hence the change.

3) Removes the extranous Assert!() call that was added by accident.

4) Removes vestigial println! calls in our testing utilites. Any prints
   that I left in where converted to panic!(), since they are a good
   indication that the test should fail.
